src/fsharp/ConstraintSolver.fs

```diff
@@ -2575,11 +2575,21 @@ let CodegenWitnessThatTypSupportsTraitConstraint tcVal g amap m (traitInfo:Trait
                 | true, true, 1 -> 
                         Some (mkStaticRecdFieldSet (rfref, tinst, argExprs.[0], m))
                 | true, false, 2 -> 
-                        Some (mkRecdFieldSet g (argExprs.[0], rfref, tinst, argExprs.[1], m))
+                        // If we resolve to an instance field on a struct and we haven't yet taken 
+                        // the address of the object then go do that 
+                        if rfref.Tycon.IsStructOrEnumTycon  && not (isByrefTy g (tyOfExpr g argExprs.[0])) then 
+                            let h = List.head argExprs
+                            let wrap,h' = mkExprAddrOfExpr g true false DefinitelyMutates h None m 
+                            Some (wrap (mkRecdFieldSetViaExprAddr (h', rfref, tinst, argExprs.[1], m)))
+                        else        
+                            Some (mkRecdFieldSetViaExprAddr (argExprs.[0], rfref, tinst, argExprs.[1], m))
                 | false, true, 0 -> 
                         Some (mkStaticRecdFieldGet (rfref, tinst, m))
                 | false, false, 1 -> 
-                        Some (mkRecdFieldGet g (argExprs.[0], rfref, tinst, m))
```

Comment:
This code is about the witness for the solution of a trait constraint when the solution is a record field getter.  I suspect there are cases where the input here could already be a byref.